handler/src/main/java/io/netty/handler/ssl/SslHandler.java

```diff
@@ -691,6 +692,24 @@ public void handlerRemoved0(ChannelHandlerContext ctx) throws Exception {
                     new ChannelException("Pending write on removal of SslHandler"));
         }
         pendingUnencryptedWrites = null;
+
+        SSLHandshakeException cause = null;
+
+        // If the handshake is not done yet we should fail the handshake promise and notify the rest of the
+        // pipeline.
+        if (!handshakePromise.isDone()) {
+            cause = new SSLHandshakeException("SslHandler removed before handshake completed");
+            if (handshakePromise.tryFailure(cause)) {
+                ctx.fireUserEventTriggered(new SslHandshakeCompletionEvent(cause));
+            }
+        }
+        if (!sslClosePromise.isDone()) {
```

Comment:
Should we be considering the case when `ctx.fireUserEventTriggered(...)` above throws and we will miss calling `notifyClosePromise(cause);`

Reply:
ctx.fireUserEventTriggered(...) will never throw by definition...